### PR TITLE
New module: add docker_stack module (cloud/docker/docker_stack)

### PR DESCRIPTION
**Pull request by dariko**
##### SUMMARY
This PR adds a new module, docker_stack, which can be used to manage docker stacks on a docker swarm. 

This module does not use the function/classes defined is [docker_common.py](/ansible/ansible/blob/devel/lib/ansible/module_utils/docker_common.py), as docker stack is not implemented in the docker API but in the docker CLI.

##### ISSUE TYPE
 - New Module Pull Request

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 5cb5228cde) last updated 2017/05/14 08:47:37 (GMT +200)
  config file = 
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/src/ansible/lib/ansible
  executable location = /home/user/src/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

###### module parameters

| parameter name | required | default | description|
|----------|:-------------:|-------|----|
|name|yes||stack name|
|state|yes||stack status (`present` or `absent`)|
|prune|no||If true will add the `--prune` option to the docker command.|
|compose_yaml|no||string containing the yaml definition of the stack
|compose_file|no||path on the module executing node of a file containing. 

###### return value
The module returns a variable `docker_stack_spec_diff` which contains
the differences between the 'Spec' field of the stack services before
and after applying the new stack definition.

For example, removing the environment variables from a service
named `test_service` in a stack named `test_stack` with this task:

```
-   name: test service
    run_once: yes
    docker_stack:
        state: present
        name: test_stack
        compose_yaml:
            version: '3'
            services:
                test_service:
                    image: alpine:latest
```

will return:
```
changed: [test_host] => {
    "changed": true,
    "docker_stack_specs_diff": "{'test_stack_test_service': {u'TaskTemplate': {u'ContainerSpec': {delete: [u'Env']}}}}",
    "invocation": {
        "module_args": {
            "compose_file": null,
            "compose_yaml": "{'services': {'test_service': {'image': 'alpine:latest'}}, 'version': '3'}",
            "name": "test_stack",
            "state": "present"
        }
    }
}
```

###### requirements
python module jsondiff

### CODE REVIEW

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        default: ""
+        description:
+        -   String containing the yaml definition of the stack
+        -   Must be in compose format
+    compose_file:
+        required: false
+        default: ""
+        description:
+        -   >
+            Path on the module executing node of a file containing
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Could you please re-phrase this, IMHO I am not able understand this.

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True),
+            'compose_yaml': dict(),
+            'compose_file': dict(),
+            'prune': dict(default=False, type='bool'),
+            'state': dict(default='present', choices=['present', 'absent'])
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: state in `DOCUMENTATION` say required but I see this is not `required=True` here.

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        supports_check_mode=False
+    )
+
+    state = module.params['state']
+    compose_yaml = module.params['compose_yaml']
+    compose_file = module.params['compose_file']
+    name = module.params['name']
+
+    if state == 'present':
+        if compose_yaml and compose_file:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: use `required_one_of=[['compose_yaml', 'compose_file']],`

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            module.fail_json(msg="both compose_file and compose_yaml " +
+                                 "parameters given")
+        elif compose_yaml:
+            fd, compose_file = mkstemp()
+            stack_file = open(compose_file, 'w')
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: use `with` statement

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            stack_file.write(compose_yaml)
+            stack_file.close()
+            os.close(fd)
+        elif not compose_file:
+            module.fail_json(msg="compose_yaml or compose_file " +
+                                 "parameters required if state==up")
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: This `state` is not mentioned in `DOCUMENTATION` or `argument_spec`

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+                             out=out,
+                             rc=rc, err=err)
+
+        before_after_differences = json_diff(before_stack_services,
+                                             after_stack_services)
+        for k in before_after_differences.keys():
+            if isinstance(before_after_differences[k], dict):
+                before_after_differences[k].pop('UpdatedAt', None)
+                before_after_differences[k].pop('Version', None)
+                if len(before_after_differences[k].keys()) == 0:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: `if not list(before_after_differences[k].keys()):`

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                    before_after_differences.pop(k)
+
+        if before_after_differences == {}:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: `if not before_after_differences:`

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                changed=True,
+                docker_stack_spec_diff=str(before_after_differences))
+
+    else:
+        rc, out, err = module.run_command(["docker",
+                                           "stack",
+                                           "down",
+                                           name])
+
+        if not rc == 0:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: `if rc != 0:`

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    ret = {}
+    for service_name in docker_stack_services(module, stack_name):
+        ret[service_name] = docker_service_inspect(module, service_name)
+    return ret
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True),
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Add type as 'str' for name

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    returned: on change
+    type: dict
+'''
+
+EXAMPLES = '''
+-   name: deploy 'stack1' stack from file
+    docker_stack:
+        state: present
+        name: stack1
+        compose_file: /opt/stack.compose
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Add blank line after each example

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+def docker_service_inspect(module, service_name):
+    rc, out, err = module.run_command(["docker",
+                                       "service",
+                                       "inspect",
+                                       service_name])
+    if rc != 0:
+        return None
+    else:
+        ret = json.loads(out)[0]['Spec']
+        return(ret)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Remove redundant parens

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+def docker_stack_services(module, stack_name):
+    rc, out, err = module.run_command(["docker",
+                                       "stack",
+                                       "services",
+                                       stack_name,
+                                       "--format",
+                                       "{{.Name}}"])
+    if out != ("Nothing found in stack %s\n" % stack_name):
+        return(out.strip().split('\n'))
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: remove redundant parens

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 8 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 7 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 1 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: you're right, now I think now it's more clear

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 10 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 12 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 11 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 9 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 2 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: removed `required: True` from docs

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 4 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: Ok, I also enclosed the whole part of code which uses the temporary file in a try-finally block, to ensure removal of the temporary file

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 5 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: It was a leftover from a previous code iteration, corrected to `if state==present`

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 6 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: done

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 3 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: Using the `required_one_of` would enforce the parameters presence even if `state==absent`, while one of them is required only if `state==present`.

**Comment by dariko** [comment]
@Akasurde I think i covered all your points.
Instead of `required_one_of` for the `compose_file` and `compose_yaml` parameters I used AnsibleModule's `mutually_exclusive` parameter, which I think it's a better fit it this situation

**Comment by ulope** [comment]
In the meantime [docker-py has gained support for swarm mode](https://docker-py.readthedocs.io/en/stable/swarm.html) - so this PR should probably be updated to use that instead of the CLI

**Comment by dariko** [comment]
@ulope sorry for the (way) late reply. I don't see anything on the docs you linked about stacks, it looks like code which can be used to join/leave/init swarms.
As far as I know `docker stack` is implemented in the docker client, not in the server..

**Comment by ulope** [comment]
@dariko You are correct that the "stack" itself is a function of the client but [services are supported in the API](https://docker-py.readthedocs.io/en/stable/services.html) which basically is all you need to build stack on top of ansible yourself.

**Comment by dariko** [comment]
@ulope I submitted #19229 some time ago, which adds a docker_service module: is this what you're looking for? It does not use `docker-py`'s service interface, which was lacking at the time.

Since I submitted this I'm not using that module anymore: this one is concise (*way* less prone to errors) and thanks to the yaml format used by `docker stack` presents a nice and clean interface.
Supporting more docker API versions in particular was growing complexity in the `docker_service` module, while this one simply delegates everything to the docker client (so the `version` attribute in the stack yaml can be used to have consistent results).

All in all I (personally) feel more confident maintaining even a single service with this module (as a one-service stack) than with the `docker_service` one, and I won't suggest using it anymore (if not for some edge cases where the `docker stack` interface might be lacking).

**Comment by gesellix** [comment]
What's missing here to get the module added to Ansible? I'm highly interested in using Docker Swarm/Stack via Ansible.

**Comment by smessmer** [comment]
+1

**Comment by ahwebd** [comment]
Got error:
`ImportError: No module named jsondiff`

**Comment by dariko** [comment]
Thanks, the `jsondiff` requirement was already explicit in the module metadata, but I added a note in this PR first message.

**Comment by chouseknecht** [comment]
Looks good to me.

+1

**Comment by jctanner** [comment]
@Akasurde can you clear your review on this?

bot_status

**Comment by LuisGerardoAngel** [comment]
Is this up for next release?

**Comment by gesellix** [comment]
@Akasurde can you please check whether your requested changes have been incorporated and clear your review?

**Comment by dariko** [comment]
@Akasurde I think I have addressed the problems you mentioned, have i missed something?

**Comment by dariko** [comment]
@Akasurde I repeat: I think I have addressed the problems you mentioned, have i missed something?

**Comment by gesellix** [comment]
@chouseknecht @bcoca maybe @Akasurde doesn't read our comments, maybe you can help to make this PR pass?

**Comment by LuisGAngel** [comment]
@Akasurde Is this up for the next release?

**Comment by Akasurde** [comment]
@LuisGerardoAngel Sorry I missed comments under the pile of notifications.

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/python
+
+# This file is part of Ansible
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Replace line 3 to 16 with 

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Copyright: (c) 2017, Ansible Project
```

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.0'}
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: s/1.0/1.1/

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+DOCUMENTATION = '''
+---
+module: docker_stack
+author: "Dario Zanzico (@dariko)"
+short_description: docker stack module
+description: |
+    Manage docker stacks using the 'docker stack' command
+    on the target node
+    (see examples)
+version_added: "2.4"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: s/2.4/2.5/

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+        -   Service state.
+        default: "present"
+        choices:
+        -   present
+        -   absent
+    compose_yaml:
+        required: false
+        default: ""
+        description:
+        -   String containing the yaml definition of the stack
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Fullstop at end of sentence.

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+
+        if not before_after_differences:
+            module.exit_json(changed=False)
+        else:
+            module.exit_json(
+                changed=True,
+                docker_stack_spec_diff=str(before_after_differences))
+
+    else:
+        rc, out, err = module.run_command(["docker",
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: Please use 

```
docker_bin = module.get_bin_path('docker', required=True)
rc, out, err = module.run_command([docker_bin, ...
```

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+                                           name])
+
+        if rc != 0:
+            module.fail_json(msg="'docker stack down' command failed",
+                             out=out,
+                             rc=rc,
+                             err=err)
+
+        module.exit_json(changed=True, msg=out, err=err)
+
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: ```if __name__ == "__main__":```

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+                                       service_name])
+    if rc != 0:
+        return None
+    else:
+        ret = json.loads(out)[0]['Spec']
+        return ret
+
+
+def docker_stack_deploy(module, stack_name, compose_file):
+    command = ["docker", "stack", "deploy"]
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: ```docker_bin = module.get_bin_path('docker', required=True)```

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+                                       stack_name,
+                                       "--format",
+                                       "{{.Name}}"])
+    if out != ("Nothing found in stack %s\n" % stack_name):
+        return out.strip().split('\n')
+    return []
+
+
+def docker_service_inspect(module, service_name):
+    rc, out, err = module.run_command(["docker",
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: ```docker_bin = module.get_bin_path('docker', required=True)```

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+'''
+
+from ansible.module_utils.basic import AnsibleModule, os
+from tempfile import mkstemp
+import json
+from jsondiff import diff as json_diff
+
+
+def docker_stack_services(module, stack_name):
+    rc, out, err = module.run_command(["docker",
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: ```docker_bin = module.get_bin_path('docker', required=True)```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Sorry missed due to lots of notifications.

**Comment by dariko** [comment]
I need some help with the errors returned by CI:
- I dot't know what this means:
  `````
  == Missing from __future__ import (absolute_import, division, print_function) ==
  ./lib/ansible/modules/cloud/docker/docker_stack.py
  `````
- There are failures related to the unavailability of the jsondiff module:
  `````
  failure: The test `ansible-test sanity --test import --python 2.6` failed with the following error: 
  lib/ansible/modules/cloud/docker/docker_stack.py:95:0: ImportError: No module named jsondiff
  `````
  but the `jsondiff` requirement is explicit in the module metadata. Should I add it somewhere else?
- Lastly, I suppose the error:
  `````
  lib/ansible/modules/cloud/docker/docker_stack.py:0:0: E321 Exception attempting to import module for argument_spec introspection
  `````
  is a consequence of the missing jsondiff module.

thanks

**Comment by mattclay** [comment]
@dariko Take a look at the help doc links. They're the `[?]` in the message posted by ansibot.

More specifically:

- For an example of a `__future__` import for a module, take a look at: https://github.com/ansible/ansible/blob/db61f8f9673bb03a34ff55e3d8d31b7225c6f47d/lib/ansible/modules/system/ping.py#L8
- For an example of handling import errors in modules, take a look at: https://github.com/ansible/ansible/blob/ba22514d479777967da292d12ac542c8c727af2e/lib/ansible/modules/source_control/gitlab_group.py#L104-L108 https://github.com/ansible/ansible/blob/ba22514d479777967da292d12ac542c8c727af2e/lib/ansible/modules/source_control/gitlab_group.py#L166-L167

**Comment by ushuz** [comment]
@dariko I've addressed some complaints of @ansibot in https://github.com/dariko/ansible/pull/2

**Comment by phoenix741** [comment]
Is it possible to add the option to activate the parameters : --with-registry-auth that i use to authenticate on a private registry ?

Thanks

**Comment by ushuz** [comment]
@phoenix741 Take a look at https://github.com/dariko/ansible/pull/2, someone has already implemented `--with-registry-auth`.

**Comment by mattclay** [comment]
@dariko Can you include integration tests for this new module?

**Comment by DimmKirr** [comment]
Is this a direct replacement of `docker_swarm_mode` module?

**Comment by dariko** [comment]
@kireevco I can't find a module named `docker_swarm_mode`.
There is a `docker_swarm` module in this PR: https://github.com/ansible/ansible/pull/38280 which seems about managing a swarm clusters and member nodes, not stacks running on the cluster.

**Comment by DimmKirr** [comment]
@dariko sorry, i meant your `docker_swarm_service` PR https://github.com/ansible/ansible/pull/19229

**Comment by dariko** [comment]
@kireevco The `docker_swarm_service` module manages a  single service while this one manages a stack (a collection of services).
I seem to remember there were some limitations on stack managed services compared to standalone services, but they might have been overcome.

**Comment by DimmKirr** [comment]
Does this module currently support (tested and is encouraged to use for) deployment of stacks to docker swarm mode?

Thanks.

**Comment by dariko** [comment]
@kireevco I haven't used it in a year or so, but I don't expect it to be broken considering there's very little to break :)
What it does is really straightforward: it executes the docker command line client on the target host using the given `compose_{yaml,file}` as parameter; validation and everything else is delegated to docker's own client.
Still, if you try it and find some problems let me know

**Comment by DimmKirr** [comment]
@dariko so bottom line, those are intended to be kept side by side? https://github.com/ansible/ansible/pull/24588 and https://github.com/ansible/ansible/pull/19229 ?

**Comment by dariko** [comment]
@kireevco there is no relation between the modules proposed in #24588 and #19229. They do not require each other.

**Comment by MadDaelim** [comment]
What with multiple compose files for docker stack deploy? Feature was added in docker 18.03.0 https://github.com/docker/cli/pull/569

**Comment by dariko** [comment]
@MadDaelim thanks, I missed that.

I think to support this feature the module interface should change: I could replace the `compose_{file,yaml}` parameters with a new one (named... `compose`? `compose_defs`?) which would be a list whose elements could be strings (treated as the old `compose_file`) or dictionaries (mapping to the old `compose_yaml`).  
ex:
`````
-   name: deploy stack1 from my_generic_compose.yml + customization
    docker_stack:
        state: present
        name: stack1
        compose:
        -   my_generic_compose.yml
        -   version: '3'
            services:
                myservice:
                    # change image
                    image: alpine:latest
        #[...] other list elements
`````
Anyone see any problem with this approach?

**Comment by dariko** [comment]
@MadDaelim I've changed the module parameters as explained in my previous comment, let me know if it works for you

**Comment by dariko** [comment]
@mattclay I'm gonna read the docs about integration testing

**Comment by MadDaelim** [comment]
@dariko looks good for me from code reading, I can't test it right now

**Review by Akasurde — APPROVED** [review]

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+        default: []
+        description:
+        -   List of compose definitions. Any element may be a string
+            referring to the path of the compose file on the target host
+            or the YAML contents of a compose file nested as dictionary.
+    prune:
+        required: false
+        default: false
+        description:
+        -   If true will add the `--prune` option to the `docker stack deploy` command.
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: You cannot use backticks for formatting here. Use `C(--prune)` and `C(docker stack deploy)` instead.

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: (There are some more places where you use backticks; I won't mark them all.)

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+    state = module.params['state']
+    compose = module.params['compose']
+    name = module.params['name']
+    absent_retries = module.params['absent_retries']
+    absent_retries_interval = module.params['absent_retries_interval']
+
+    if state == 'present':
+        if not compose:
+            module.fail_json(msg=("compose parameter must be a list "
+                                  "containing at least an element"))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: maybe 'an' -> 'one'?

File: lib/ansible/modules/cloud/docker/docker_stack.py [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+                                     "or a dictionary" % compose_def)
+
+            before_stack_services = docker_stack_inspect(module, name)
+
+            rc, out, err = docker_stack_deploy(module, name, compose_files)
+
+            after_stack_services = docker_stack_inspect(module, name)
+
+        finally:
+            for temp_file in temp_files:
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_stack.py]: `finally` blocks are not executed if `module.fail_json` is used (`fail_json()` calls `sys.exit()`)!

To make sure that a file is deleted, use `module.add_cleanup_file(filename)` after creating it.

File: test/integration/targets/docker_stack/tasks/test_stack.yml [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+  register: output
+  docker_stack:
+    state: absent
+    name: test_stack
+    absent_retries: 30
+
+- name: assert state=absent idempotency
+  assert:
+    that:
+    - not output is changed
```
Comment (felixfontein) [inline test/integration/targets/docker_stack/tasks/test_stack.yml]: why not `output is not changed`?

**Review by felixfontein — COMMENTED** [review]
I've never used `docker stack`, so I can't comment on the main part of the module, but I found some little things (and one not so little) while looking over it :)

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 22 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: Thanks, I'm changing them

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 23 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: That was the idea... I'm changing it

File: lib/ansible/modules/cloud/docker/docker_stack.py (hunk 24 above)
Comment (dariko) [reply lib/ansible/modules/cloud/docker/docker_stack.py]: Ok, I removed the whole try..finally: using `module.add_cleanup_file` it is now unnecessary

File: test/integration/targets/docker_stack/tasks/test_stack.yml (hunk 25 above)
Comment (dariko) [reply test/integration/targets/docker_stack/tasks/test_stack.yml]: This test is verifying the module does **not** report the task as changed when run with `state==absent` and a name referring to a stack which does not exist (in this case, it should have been just deleted).

Comment (felixfontein) [reply test/integration/targets/docker_stack/tasks/test_stack.yml]: `not A is B` is equivalent to `A is not B`. I find the second form easier to read, as it is more like natural language.

Comment (dariko) [reply test/integration/targets/docker_stack/tasks/test_stack.yml]: Sorry, I misunderstood your previous comment.  
I changed the test as you suggested.

Comment (felixfontein) [reply test/integration/targets/docker_stack/tasks/test_stack.yml]: No problem :) Thanks!

**Review by felixfontein — APPROVED** [review]
shipit

**Review by jwitko — APPROVED** [review]

**Comment by jwitko** [comment]
shipit

**Review by tbouvet — APPROVED** [review]

**Comment by tbouvet** [comment]
shipit

**Comment by gundalow** [comment]
Merged into `devel` for release in Ansible 2.8.
Thank you all for the reviews

If you search for `docker` on https://ansible.sivel.net/pr/byfile.html there are a few other PRs open. Reviews on those would be very welcome.

If you'd like to become maintainers of Docker (so you are notified of issues & PR) and have voting rights, we'd welcome you to raise a PR to add yourself to https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml

Any questions feel free to ping me on `#ansible-devel`

**Comment by mattclay** [comment]
The tests in this PR are conflicting with the tests for docker_swarm:

https://app.shippable.com/github/ansible/ansible/runs/84464/50/tests

I'm looking into the issue.

**Comment by mattclay** [comment]
Fix for the issue: https://github.com/ansible/ansible/pull/45807

**Comment by mattclay** [comment]
The "fix" has been merged. It's really a work-around at this point which will hopefully resolve the issues with tests failing.

I ended up adding a task to the end of the docker_stack test to leave the swarm. This should prevent future docker tests from having issues. However, doing this exposed another problem: it wasn't possible to leave the swarm if the stack had been updated. I don't know why this happens, so I ended up disabling that part of the test so the rest of the test could still execute.

Hopefully someone has time to investigate further and fix the underlying issue.

**Pull request merged**